xds/src/main/java/io/grpc/xds/LoadStatsManager2.java

```diff
@@ -0,0 +1,396 @@
+          ReferenceCounted<ClusterLocalityStats> ref = localityStats.get(locality);
+          ClusterLocalityStatsSnapshot snapshot = ref.get().snapshot();
+          // Only discard stats object after all in-flight calls under recording had finished.
+          if (ref.getReferenceCount() == 0 && snapshot.callsInProgress == 0) {
+            localitiesToDiscard.add(locality);
+          }
+          UpstreamLocalityStats.Builder localityStatsBuilder = UpstreamLocalityStats.newBuilder();
+          localityStatsBuilder.setLocality(locality);
+          localityStatsBuilder.setTotalIssuedRequests(snapshot.callsIssued);
+          localityStatsBuilder.setTotalSuccessfulRequests(snapshot.callsSucceeded);
+          localityStatsBuilder.setTotalErrorRequests(snapshot.callsFailed);
+          localityStatsBuilder.setTotalRequestsInProgress(snapshot.callsInProgress);
+          builder.addUpstreamLocalityStats(localityStatsBuilder.build());
+          // Use the max (drops/loads) recording interval as the overall interval for the
+          // cluster's stats. In general, they should be mostly identical.
+          builder.setLoadReportIntervalNanos(
+              Math.max(builder.getLoadReportIntervalNanos(), snapshot.durationNano));
+        }
+        localityStats.keySet().removeAll(localitiesToDiscard);
+        if (localityStats.isEmpty()) {
+          toDiscard.add(edsServiceName);
+        }
+      }
+      clusterLoadStats.keySet().removeAll(toDiscard);
+    }
+    List<ClusterStats> res = new ArrayList<>();
+    for (ClusterStats.Builder builder : statsReportBuilders.values()) {
+      res.add(builder.build());
+    }
+    return Collections.unmodifiableList(res);
+  }
+
+  /**
+   * Gets the traffic stats (drops and loads) as a list of {@link ClusterStats} recorded for all
+   * clusters since the previous call of this method or {@link #getClusterStatsReports} for each
+   * specific cluster. A {@link ClusterStats} includes stats for a specific cluster with
+   * edsServiceName.
+   */
+  synchronized List<ClusterStats> getAllClusterStatsReports() {
+    Set<String> allClusters = Sets.union(allDropStats.keySet(), allLoadStats.keySet());
+    List<ClusterStats> res = new ArrayList<>();
+    for (String cluster : allClusters) {
+      res.addAll(getClusterStatsReports(cluster));
+    }
+    return Collections.unmodifiableList(res);
+  }
+
+  /**
+   * Recorder for dropped requests. One instance per cluster with edsServiceName.
+   */
+  @ThreadSafe
+  class ClusterDropStats {
+    private final String clusterName;
+    @Nullable
+    private final String edsServiceName;
+    private final AtomicLong uncategorizedDrops = new AtomicLong();
+    private final ConcurrentMap<String, AtomicLong> categorizedDrops = new ConcurrentHashMap<>();
+    private final Stopwatch stopwatch;
+
+    private ClusterDropStats(
+        String clusterName, @Nullable String edsServiceName, Stopwatch stopwatch) {
```

Comment:
Now ClusterDropStats creation is LoadStatsManager2 internal, the outer class has the full control for Stopwatch instances to be passed in.